internal/pkg/docker/dockerengine/dockerengine.go

```diff
@@ -246,11 +258,51 @@ func (in *RunOptions) generateRunArguments() []string {
 
 // Run runs a Docker container with the sepcified options.
 func (c DockerCmdClient) Run(ctx context.Context, options *RunOptions) error {
-	//Execute the Docker run command.
-	if err := c.runner.RunWithContext(ctx, "docker", options.generateRunArguments()); err != nil {
-		return fmt.Errorf("running container: %w", err)
+	// set default options
+	if options.LogOptions.Color == nil {
+		options.LogOptions.Color = color.New()
+	}
+	if options.LogOptions.Output == nil {
+		options.LogOptions.Output = os.Stderr
+	}
+
+	// Ensure only one thread is writing to Output at a time
+	// since we don't know if the Writer is thread safe.
+	mu := &sync.Mutex{}
+	g, ctx := errgroup.WithContext(ctx)
+	logger := func() (io.Writer, func()) {
```

Comment:
stupid question: why not 
```suggestion
	logger := func() *io.PipeWriter {
```

Reply:
great question!! i try to answer it down on line 291 - but basically `scanner.Scan()` will only return if `pw` is closed. (when `pw` is closed, `pr` gets an EOF, which causes `scanner.Scan()` to return false). 

in the "real" code, where we use `os/exec.Cmd`, `Wait()`/`Run()` returning will return EOF to the readers when the command exits, causing `scanner.Scan()` to finish. however, in tests, we don't close the reader so the logger routines block forever until the test times out.

Reply:
oh, i realized your question was to just return the pipe writer instead lol. that's actually a good idea :joy: